src/otaclient/_logging.py

```diff
@@ -19,32 +19,181 @@
 import atexit
 import contextlib
 import logging
+import time
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
 from queue import Queue
 from threading import Event, Thread
-from urllib.parse import urljoin
+from urllib.parse import urljoin, urlparse
 
+import grpc
 import requests
+from otaclient_iot_logging_server_pb2.v1 import (
+    otaclient_iot_logging_server_v1_pb2 as log_pb2,
+)
+from otaclient_iot_logging_server_pb2.v1 import (
+    otaclient_iot_logging_server_v1_pb2_grpc as log_v1_grpc,
+)
+from pydantic import AnyHttpUrl
 
 from otaclient.configs.cfg import cfg, ecu_info, proxy_info
 
 
+class LogType(Enum):
+    LOG = "log"
+    METRICS = "metrics"
+
+
+@dataclass
+class QueueData:
+    """Queue data format for logging."""
+
+    log_type: LogType
+    message: str
+
+
+class Transmitter(ABC):
+    def __init__(self, logging_upload_endpoint, ecu_id: str):
+        self.ecu_id = ecu_id
+
+    @abstractmethod
+    def send(self, log_type: LogType, message: str, timeout: int) -> None:
+        pass
+
+    @abstractmethod
+    def check(self, timeout: int) -> None:
+        pass
+
+
+class TransmitterGrpc(Transmitter):
+    def __init__(self, logging_upload_grpc_endpoint: AnyHttpUrl, ecu_id: str):
+        super().__init__(logging_upload_grpc_endpoint, ecu_id)
+
+        parsed_url = urlparse(str(logging_upload_grpc_endpoint))
+        log_upload_endpoint = parsed_url.netloc
+        channel = grpc.insecure_channel(log_upload_endpoint)
+        self._stub = log_v1_grpc.OTAClientIoTLoggingServiceStub(channel)
+
+    def send(self, log_type: LogType, message: str, timeout: int) -> None:
+        pb2_log_type = (
+            log_pb2.LogType.METRICS
+            if log_type == LogType.METRICS
+            else log_pb2.LogType.LOG
+        )
```

Comment:
Inline if else here is OK, but implementing a `convert` method on `otaclient._logging.LogType` Enum is also a good idea.

Reply:
Good catch. Fixed.